File: clojure/beer-song/src/beer_song.clj
(ns beer-song
  (use [clojure.string])
)

(defn- verse-0 []
  (str 
    "No more bottles of beer on the wall, no more bottles of beer.\n"
    "Go to the store and buy some more, 99 bottles of beer on the wall.\n"
  )
)

(defn- verse-1 []
  (str 
    "1 bottle of beer on the wall, 1 bottle of beer.\n"
    "Take it down and pass it around, no more bottles of beer on the wall.\n"
  )
)

(defn- verse-2 []
  (str 
    "2 bottles of beer on the wall, 2 bottles of beer.\n"
    "Take one down and pass it around, 1 bottle of beer on the wall.\n"
  )
)

(defn verse [n]
  (cond
    (= n 0) (verse-0)
    (= n 1) (verse-1)
    (= n 2) (verse-2)
    :else (str 
      n " bottles of beer on the wall, " n " bottles of beer.\n"
      "Take one down and pass it around, " (dec n) " bottles of beer on the wall.\n"
    )
  )
)

(defn- sing-range [list]
  (join "\n" 
    (map verse list)
  )
)

(defn sing 
  ([from] (sing from 0))
  ([from to] (sing-range (range from (dec to) -1)))
)

File: fsharp/beer-song/BeerSong.fs
module BeerSong

let capital (text: string) =
    text.[..0].ToUpper() + text.[1..]

let bottles (n: int) =
    match n with
    | -1 -> "99 bottles of beer"
    | 0 -> "no more bottles of beer"
    | 1 -> "1 bottle of beer"
    | _ -> string n + " bottles of beer"

let action (n: int) = 
    match n with
    | 0 -> "Go to the store and buy some more"
    | 1 -> "Take it down and pass it around"
    | _ -> "Take one down and pass it around"

let verse (n: int) = 
    (bottles n |> capital) + " on the wall, " + 
    bottles n + ".\n" + 
    action n + ", " +
    bottles(n-1) + " on the wall.\n"

let verses (stop: int) (start: int): string = 
    ( [start..stop] 
        |> List.rev 
        |> List.map verse 
        |> String.concat "\n"
    ) + "\n"    

let sing: string = 
    verses 99 0